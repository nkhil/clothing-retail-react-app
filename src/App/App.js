import React, { Component, Fragment } from "react"
import products from "./helpers/products"
import voucherCodes from "./helpers/voucherCodes"
import ProductList from "./components/ProductList"
import ShoppingCart from "./components/Cart"

class App extends Component {
  state = {
    products: {},
    shoppingCart: {},
    voucherCodes: [],
    activeVoucherCode: null,
  }

  componentWillMount = () => {
    this.loadProductsIntoState(products)
    this.loadVoucherCodesIntoState(voucherCodes)
  }

  loadProductsIntoState = products => {
    this.setState({ products })
  }

  loadVoucherCodesIntoState = voucherCodes => {
    this.setState({ voucherCodes })
  }

  addToCart = key => {
    const shoppingCart = { ...this.state.shoppingCart }
    shoppingCart[key] = shoppingCart[key] + 1 || 1
    this.setState({ shoppingCart })
  }

  removeFromCart = key => {
    const shoppingCart = { ...this.state.shoppingCart }
    shoppingCart[key] = shoppingCart[key] - 1
    if (shoppingCart[key] <= 0) {
      delete shoppingCart[key]
    }
    this.setState({ shoppingCart })
  }

  deductProductFromInventory = key => {
    let products = { ...this.state.products }
    products[key].productQuantity -= 1
    this.setState({ products })
  }

  calculateTotal = () => {
    const cartIds = Object.keys(this.state.shoppingCart)
    const cartTotal = cartIds.reduce((prevTotal, key) => {
      const productPrice = this.state.products[key].productPrice
      const count = this.state.shoppingCart[key]
      return prevTotal + productPrice * count
    }, 0)
    return cartTotal
  }

  calculateDiscountedTotal = () => {
    if (this.state.activeVoucherCode) {
      const discountAmount = this._getVoucherObject(
        this.state.activeVoucherCode
      ).discountAmount
      const discountedTotal = this.calculateTotal() - discountAmount
      return discountedTotal
    }
    return this.calculateTotal()
  }

  setActiveVoucherCode = voucherCode => {
    this.setState({ activeVoucherCode: voucherCode })
  }

  voucherCodeIsValid = voucherCode => {
    const voucherObject = this._getVoucherObject(voucherCode)
    if (voucherObject && voucherCode === "FIFTEENOFF") {
      return this._checkFifteenOffCriteria(voucherObject.minimumSpend)
    }
    if (voucherObject) {
      return this._voucherMeetsMinimumSpend(voucherObject.minimumSpend)
    }
    return false
  }

  _checkFifteenOffCriteria = minimumSpend => {
    return (
      this._voucherMeetsMinimumSpend(minimumSpend) &&
      this._shoppingCartContainsFootwear()
    )
  }

  _shoppingCartContainsFootwear = () => {
    const shoppingCartIds = Object.keys(this.state.shoppingCart)
    const footwearItemsInCart = shoppingCartIds.filter(key => {
      return this.state.products[key].productCategory.includes("Footwear")
    })
    return footwearItemsInCart.length > 0
  }

  _voucherMeetsMinimumSpend = minimumSpend =>
    this.calculateDiscountedTotal() >= minimumSpend

  _getVoucherObject = voucherCode => {
    const voucherObject = this.state.voucherCodes.filter(
      voucher => voucher.code === voucherCode
    )
    return voucherObject[0]
  }

  render = () => {
    return (
      <Fragment>
        <ProductList
          products={this.state.products}
          addToCart={this.addToCart}
          deductProductFromInventory={this.deductProductFromInventory}
        />
        <ShoppingCart
          products={this.state.products}
          shoppingCart={this.state.shoppingCart}
          removeFromCart={this.removeFromCart}
          calculateDiscountedTotal={this.calculateDiscountedTotal}
          voucherCodeIsValid={this.voucherCodeIsValid}
          setActiveVoucherCode={this.setActiveVoucherCode}
        />
      </Fragment>
    )
  }
}

export default App
