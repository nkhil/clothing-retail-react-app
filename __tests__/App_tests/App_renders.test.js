import "@babel/polyfill"
import React from "react"
import App from "../../src/App/App"
import ReactDOM from "react-dom"
import renderer from "react-test-renderer"

require("react-router-dom")
jest.mock("react-router-dom")

test("App renders without crashing", () => {
  const div = document.createElement("div")
  ReactDOM.render(<App />, div)
})

test("App renders correctly", () => {
  const app = renderer.create(<App />).toJSON()
  expect(app).toMatchSnapshot()
})
