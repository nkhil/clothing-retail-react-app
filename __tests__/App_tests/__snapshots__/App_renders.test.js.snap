// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div
  className="container"
>
  <ul
    className="products-list"
  >
    <li
      className="product"
      id="product1"
    >
      <div
        className="product-details"
      >
        <h3>
          Almond Toe Court Shoes, Patent Black
        </h3>
        <h4>
          £99.00
        </h4>
        <p>
          5 available
        </p>
        <p>
          Category: 
          Women's Footwear
        </p>
      </div>
      <button
        className="product1-button item-button"
        disabled={false}
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
    <li
      className="product"
      id="product2"
    >
      <div
        className="product-details"
      >
        <h3>
          Suede Shoes, Blue
        </h3>
        <h4>
          £42.00
        </h4>
        <p>
          4 available
        </p>
        <p>
          Category: 
          Women's Footwear
        </p>
      </div>
      <button
        className="product2-button item-button"
        disabled={false}
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
    <li
      className="product"
      id="product3"
    >
      <div
        className="product-details"
      >
        <h3>
          Leather Driver Saddle Loafers, Tan
        </h3>
        <h4>
          £34.00
        </h4>
        <p>
          12 available
        </p>
        <p>
          Category: 
          Men's Footwear
        </p>
      </div>
      <button
        className="product3-button item-button"
        disabled={false}
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
    <li
      className="product"
      id="product4"
    >
      <div
        className="product-details"
      >
        <h3>
          Flip Flops, Red
        </h3>
        <h4>
          £19.00
        </h4>
        <p>
          6 available
        </p>
        <p>
          Category: 
          Men's Footwear
        </p>
      </div>
      <button
        className="product4-button item-button"
        disabled={false}
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
    <li
      className="product"
      id="product5"
    >
      <div
        className="product-details"
      >
        <h3>
          Flip Flops, Blue
        </h3>
        <h4>
          £19.00
        </h4>
        <p>
          
        </p>
        <p>
          Category: 
          Men's Footwear
        </p>
      </div>
      <button
        className="product5-button item-button"
        disabled={true}
        onClick={[Function]}
      >
        Sold out!
      </button>
    </li>
    <li
      className="product"
      id="product6"
    >
      <div
        className="product-details"
      >
        <h3>
          Gold Button Cardigan, Black
        </h3>
        <h4>
          £167.00
        </h4>
        <p>
          6 available
        </p>
        <p>
          Category: 
          Women’s Casualwear
        </p>
      </div>
      <button
        className="product6-button item-button"
        disabled={false}
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
    <li
      className="product"
      id="product7"
    >
      <div
        className="product-details"
      >
        <h3>
          Cotton Shorts, Medium Red
        </h3>
        <h4>
          £30.00
        </h4>
        <p>
          5 available
        </p>
        <p>
          Category: 
          Women’s Casualwear
        </p>
      </div>
      <button
        className="product7-button item-button"
        disabled={false}
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
    <li
      className="product"
      id="product8"
    >
      <div
        className="product-details"
      >
        <h3>
          Fine Stripe Short Sleeve Shirt, Grey
        </h3>
        <h4>
          £49.99
        </h4>
        <p>
          9 available
        </p>
        <p>
          Category: 
          Men’s Casualwear
        </p>
      </div>
      <button
        className="product8-button item-button"
        disabled={false}
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
    <li
      className="product"
      id="product9"
    >
      <div
        className="product-details"
      >
        <h3>
          Fine Stripe Short Sleeve Shirt, Green
        </h3>
        <h4>
          £39.99
        </h4>
        <p>
          9 available
        </p>
        <p>
          Category: 
          Men’s Casualwear
        </p>
      </div>
      <button
        className="product9-button item-button"
        disabled={false}
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
    <li
      className="product"
      id="product10"
    >
      <div
        className="product-details"
      >
        <h3>
          Sharkskin Waistcoat, Charcoal
        </h3>
        <h4>
          £75.00
        </h4>
        <p>
          2 available
        </p>
        <p>
          Category: 
          Men’s Formalwear
        </p>
      </div>
      <button
        className="product10-button item-button"
        disabled={false}
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
    <li
      className="product"
      id="product11"
    >
      <div
        className="product-details"
      >
        <h3>
          Bird Print Dress, Black
        </h3>
        <h4>
          £270.00
        </h4>
        <p>
          10 available
        </p>
        <p>
          Category: 
          Women’s Formalwear
        </p>
      </div>
      <button
        className="product11-button item-button"
        disabled={false}
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
    <li
      className="product"
      id="product12"
    >
      <div
        className="product-details"
      >
        <h3>
          Mid Twist Cut­Out Dress, Pink
        </h3>
        <h4>
          £540.00
        </h4>
        <p>
          5 available
        </p>
        <p>
          Category: 
          Women’s Formalwear
        </p>
      </div>
      <button
        className="product12-button item-button"
        disabled={false}
        onClick={[Function]}
      >
        Add to cart
      </button>
    </li>
  </ul>
  <div
    className="shopping-cart-container"
  >
    <h2>
      Shopping Cart:
    </h2>
    <ul
      className="shopping-cart-list"
    />
    <form
      id="voucher-form"
      onSubmit={[Function]}
    >
      <input
        name="voucher-code"
        placeholder="Voucher code"
        type="text"
      />
      <button
        type="submit"
      >
        Apply
      </button>
    </form>
    <h3
      id="total"
    >
      Total: 
      £0.00
    </h3>
  </div>
</div>
`;
